src/postamble.js

```diff
@@ -202,10 +202,11 @@ function run() {
 #endif
 
 #if HAS_MAIN
+    {{{ makeModuleReceiveWithVar('noInitialRun', undefined, !INVOKE_RUN) }}}
 #if MAIN_READS_PARAMS
-    if (shouldRunNow) callMain(args);
+    if (!noInitialRun) callMain(args);
```

Comment:
Ah but I guess this change already fixes this problem for me because this new code assigns `noInitialRun` directly before reading it.

Reply:
> It would be helpful if we could check for `Module['noInitialRun']` here rather than copying it into a variable. I have a use case with memory snapshots where if we have a memory snapshot we want to not run main. But I want to get the other Emscripten setup logic going while waiting on the snapshot network requests. So I use `addRunDependency` and then when the request comes back if I get a snapshot I set `Module.noInitialRun` to `true` and do other setup. I need to patch this line for this to work.

While I'm glad that works, that doesn't sound like something that emscripten officially supporteds.  I'm glad this changes makes it better, but I'm afraid we don't make any promises that will continue to work in the future.

Reply:
Is there a reason not to allow setting `Module.noInitialRun` in `preRun` hooks? It seems to me that as long as it is set ahead of `callMain()` it ought to work. At least as long as all it does is suppress this one call.

Reply:
Anyways we use a lot of emscripten implementation details and certainly don't expect the upgrade process to be simple. The most common problems are weird llvm regressions in compilation of C++ code.

Reply:
Yeah it's helpful.